### PR TITLE
Flag a course

### PR DESCRIPTION
**Pull request by b3njamint**
### Current Status 

Can flag and unflag course at free will. Changing pages and jumping around in history will not cause problems.

### Summary 

This pull request allows LSC to flag a course when there is a "bad situation," or when there are potential matches for a course but they are not ideal (i.e. there is only one grad student and one freshmen waiting to be matched. Such a flag would remind LSC to not click the match button.

- [x] implemented X
- [ ] fixed Y

### Test Plan 

The dashboard. You can see the little bookmark icons (could be changed later).
Screen Shot 2023-04-16 at 9 15 59 PM
The emulator. As you can see, there is no flag field right now. Note that this is important, because even if there is no flag field, there are no null problems.
Screen Shot 2023-04-16 at 9 16 32 PM
Oh no, a bad situation!
Screen Shot 2023-04-16 at 9 17 05 PM
Let's flag it.
Screen Shot 2023-04-16 at 9 17 14 PM
A new field is created in the Firebase, and set accordingly. We can refresh, move back and forth in history and the bookmark will still be in its correct state.
Screen Shot 2023-04-16 at 9 20 18 PM

### Notes 

Next steps (look out for a future PR):
- [ ] remove flag when new student joins
- [ ] maybe add some filter option to filter for flags?

### Breaking Changes 

- Changed the CourseContext (added a field to Firebase). I don't see any problems through testing but please let me know.

### CODE REVIEW

**Review by rgu0114 — COMMENTED** [review]
Nice work Benji! I like how you separated out all your functions to flag a course and used defensive programming to set the flagged field of CourseCard. Everything looks good for the most part, but it would be great if you could describe your testing plan a bit more.

**Comment by b3njamint** [comment]
> Nice work Benji! I like how you separated out all your functions to flag a course and used defensive programming to set the flagged field of CourseCard. Everything looks good for the most part, but it would be great if you could describe your testing plan a bit more.

Thanks for the feedback. I haven't updated the PR description yet, but will do that right now.

**Review by mluo24 — APPROVED** [review]
Good job Benji! I like your integration with the CourseContext and the elegant solution to this! I played around with it a bit and it works well.

I'm looking forward to your second PR :)

File: backend/functions/src/course/routes.ts [hunk 1]
```diff
@@ -42,6 +43,15 @@ router.get('/students/:courseId', (req, res) => {
...
 
+// Updates the flagged status in the database
+router.post('/flagged', (req, res) => {
+  const courseId = req.body.courseId
+  const flag = req.body.flagged
+  setFlagged(courseId, flag)
+    .then((data) => res.status(200).send({ flagged: flag }))
+    .catch((err) => {
+      logger.error(`Unexpected error updating flagged status: ${err.message}`)
+      res.status(500).send({ message: err.message })
```
Comment (mluo24) [inline backend/functions/src/course/routes.ts:54]: You're missing braces and parentheses at the end.

**Review by mluo24 — CHANGES_REQUESTED** [review]
wait I forgot to say that something went wrong when I ran it because it was a syntax error. I pointed it out in the code.

**Review by mluo24 — APPROVED** [review]
works now! good job!

**Pull request merged**